### PR TITLE
Opal 565 containerize

### PR DESCRIPTION
**Pull request by CK-USAF**
PULL OPAL-565-Containerize BRANCH
```
git clone https://github.com/AFMC-MAJCOM/task-queue.git
git checkout OPAL-565-Containerize
```

BUILD TASK QUEUE IMAGE
```
docker build . -t task-queue 
```

SPIN UP POSTGRES SERVER
```
docker pull postgres 
docker run --name my-postgres -e POSTGRES_PASSWORD=my_password -d -p 5432:5432 postgres
```

RUN SERVER

_Ensure on this step that the server is continuously running after the docker container 
is spun up (ie the terminal where you ran that command should be unusable until you hit ctrl c)_

1. make an env file 'env.list' with the following 
```
QUEUE_IMPLEMENTATION=sql-json
SQL_QUEUE_CONNECTION_STRING=postgresql://postgres:my_password@host.docker.internal:5432/postgres
SQL_QUEUE_NAME=test
```
2. for linux users
```
docker run -it --add-host=host.docker.internal:host-gateway --rm -p 8001:80 --env-file ./env.list task-queue server 
```

2. for windows users
```
docker run -it --rm -p 8001:80 --env-file ./env.list task-queue server 
```

RUN CONTROLLER
```
docker run -it --rm task-queue controller --help
```

### CODE REVIEW

**Review by AM-USAF — DISMISSED** [review]
Worked for me on Linux

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Ignore all hidden git files
```
Comment (JR-USAF) [inline .dockerignore:1]: Add details on how to build and run the image in both controller mode and server mode.

File: Dockerfile
Comment (JP-USAF) [inline Dockerfile:1]: Consider setting an unprivileged user in the docker file to prevent privilege escalation attacks.

**Review by JP-USAF — DISMISSED** [review]

File: README.md [hunk 2]
```diff
@@ -74,3 +74,65 @@ The `work_queue_service_cli.py` file will run a persistent service that periodic
...
+
+## Downloading Docker Image
+
+To download the docker image run the following command.
+
+```
+docker pull ghcr.io/afmc-majcom/task-queue/task-queue:latest
+```
+
+To verify this worked run. You should now see the task queue image listed out.
```
Comment (AM-USAF) [inline README.md:90]: This is a little confusing when reading, maybe add "To verify this worked, run the following command. Then, you cshould see..." or just separate the two lines with that command

**Review by AM-USAF — APPROVED** [review]

File: Dockerfile
Comment (AM-USAF) [reply Dockerfile:1]: @JP-USAF Can you explain what you mean, I'm not very familiar with docker

Comment (JP-USAF) [reply Dockerfile:1]: Docker commands run as root by default unless another user is specified. This link explains it more in depth, https://www.docker.com/blog/understanding-the-docker-user-instruction/

OWASP suggests setting the user with the following commands. (https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html)

RUN groupadd -r myuser && useradd -r -g myuser myuser
USER myuser

Comment (CK-USAF) [reply Dockerfile:1]: I added a default user to the Dockerfile

**Review by AM-USAF — DISMISSED** [review]

File: README.md [hunk 3]
```diff
@@ -74,3 +74,65 @@ The `work_queue_service_cli.py` file will run a persistent service that periodic
...
+
+You can modify the ports, passwords, and name to what you would like them to be but make sure the next steps reflect those changes.
+
+### Running the Server
+
+Before you can do this step you need to have a postgres server running and access to the server. Once you have that you can create the env.list file. The env.list file requires the following values.
+
+```
+QUEUE_IMPLEMENTATION=sql-json
+SQL_QUEUE_CONNECTION_STRING=<postgresql://postgres:your_password@hostname:port/database_name>
```
Comment (JR-USAF) [inline README.md:115]: Can we add an example of what the connection string would be, given the ports shown in the README

postgresql://postgres:<your_password>@host.docker.internal:5432/postgres

File: README.md [hunk 4]
```diff
@@ -74,3 +74,65 @@ The `work_queue_service_cli.py` file will run a persistent service that periodic
...
+docker run --add-host=host.docker.internal:host-gateway --rm -p 8001:80 --env-file ./env.list task-queue server 
+```
+
+### Running the CLI
+
+Run the following command to output help from CLI
+
+```
+docker run --rm task-queue controller --help
+```
```
Comment (JR-USAF) [inline README.md:137]: Add the example of running the server as well.

File: README.md [hunk 5]
```diff
@@ -74,3 +74,65 @@ The `work_queue_service_cli.py` file will run a persistent service that periodic
...
+QUEUE_IMPLEMENTATION=sql-json
+SQL_QUEUE_CONNECTION_STRING=<postgresql://postgres:your_password@hostname:port/database_name>
+SQL_QUEUE_NAME=<the_name_of_your_queue>
+```
+
+Now you can run the server using the following command. Note, if you started your own local postgres server your hostname should be _host.docker.internal_
+
+```
+docker run --rm -p 8001:80 --env-file ./env.list task-queue server 
+```
```
Comment (JR-USAF) [inline README.md:123]: I don't know if this is the case or not. But is the image name `task-queue` supposed to be consistent with the docker pull command `ghcr.io/afmc-majcom/task-queue/task-queue:latest`

**Review by CW-USAF — DISMISSED** [review]

**Review by JR-USAF — APPROVED** [review]

**Review by JP-USAF — APPROVED** [review]

**Review by AM-USAF — APPROVED** [review]

**Pull request merged**